iree/compiler/Dialect/HAL/Transforms/PackAllocations.cpp

```diff
@@ -0,0 +1,367 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <list>
+#include <memory>
+#include <utility>
+
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/HAL/Target/TargetBackend.h"
+#include "iree/compiler/Dialect/HAL/Target/TargetRegistry.h"
+#include "iree/compiler/Dialect/HAL/Transforms/Passes.h"
+#include "iree/compiler/Dialect/HAL/Utils/TypeUtils.h"
+#include "llvm/ADT/MapVector.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace HAL {
+
+class PackAllocationsPass
+    : public PassWrapper<PackAllocationsPass, OperationPass<FuncOp>> {
+ public:
+  using Slice = IREE::HAL::AllocatorPackOp::Slice;
+
+  explicit PackAllocationsPass(TargetOptions targetOptions)
+      : targetOptions_(targetOptions) {}
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<mlir::StandardOpsDialect>();
+    registry.insert<IREE::HAL::HALDialect>();
+  }
+
+  void runOnOperation() override {
+    auto funcOp = getOperation();
+
+    // Derive buffer constraints based on target backends.
+    // TODO(benvanik): move to a module-level attribute so that we can query
+    // this without having to plumb through the target options and share code
+    // with IdentifyConstantPools.
+    auto bufferConstraints = computeConservativeBufferConstraints(
+        targetOptions_, funcOp.getContext());
+    if (!bufferConstraints) {
+      funcOp.emitWarning() << "no target backends provided buffer "
+                              "constraints; falling back to host default";
+      bufferConstraints =
+          TargetBackend::makeDefaultBufferConstraints(funcOp.getContext());
+    }
+
+    // NOTE: we could try several algorithms and compute which packs best. For
+    // now we just pack greedily as it's fast and what most existing ML
+    // frameworks do.
```

Comment:
Could also, particularly for static cases, use a search algorithm (read: ML magic) here. We can get away with some crazy things when this runs during offline / ahead-of-time compilation :)

Reply:
yep! Profile-guided feedback would also be good and could help the dynamic cases - if we can encode that in the IR that's best: if we know for example that dimensions 0 of a `tensor<?x1x8xf32>` is always less than dimension 2 of `tensor<8x4x?xf32>` we would know that the first would always be <= size of the second and could bucket them smarter. We can also do this optimistically: emit the code for the fully dynamic size-equality packing *and* the optimistic packing where certain sizes are equal and then choose at runtime as all of that would end up as a few `select` ops. Lots of fun approaches :)